### ISSUE TITLE
VSCode gave me a "Bad text encoding" error

### ISSUE THREAD

**Issue by beibeibeibei2022** [idleberg/vscode-nsis#46]
## Summary
UTF8WithoutBOM support. I have some violent code maybe fix it.

## Details
Hi, this is my first time to submit a issue.
After I read #31, I fix the compile error, the generated installation package can run correctly.
(fix error like this screenshot)
![image](https://user-images.githubusercontent.com/98100055/213911769-b911cf4f-bc21-4a37-a3fa-1da3941e3efa.png)

But in the VSCode , I still got a red line error in VSCode.
By the way, I think #31 was wrong, UTF8withBOM works perfectly, and **UTF8withoutBOM is trouble**.

My screenshot
![图像 003](https://user-images.githubusercontent.com/98100055/213908714-aa09f10b-89dc-4085-bce3-45845d522e70.png)
#31 screenshot
![image](https://user-images.githubusercontent.com/98100055/213910247-41855daa-051f-49d8-befb-a2d7a1a52c75.png)

perhaps NSIS knows that I am using a UTF-8 format script file without BOM, so the compile is normal.
But the vscode-nsis extension doesn't know, so vscode-nsis asked vscode to report an error, and then I got the error.

## Code
I have this, Can this solve the problem?
**Original code**(TypeScript)
```ts
// https://github.com/idleberg/vscode-nsis/raw/main/src/util.ts
function findErrors(input: string): unknown {
    const result = /(?<message>.*)\r?\n.*rror in script:? "(?<file>.*)" on line (?<line>\d+)/.exec(input);

    if (result !== null) {
        const errorLine = parseInt(result.groups.line) - 1;

        return {
            code: '',
            message: result.groups.message,
            range: new vscode.Range(new vscode.Position(errorLine, 0), new vscode.Position(errorLine, getLineLength(errorLine))),
            severity: vscode.DiagnosticSeverity.Error
        };
    }

    return {};
}
```
**Modify on the basis of %USERPROFILE%\.vscode\extensions\idleberg.nsis-3.52.0\lib\extension.js**(JavaScript)
```js
let firstVdefaultwindowshowInformationMessage = true;
function findErrors(input) {
    const result = /(?<message>.*)\r?\n.*rror in script:? "(?<file>.*)" on line (?<line>\d+)/.exec(input);

    // -*- START -*-
    // by translation software
    // V.default.workspace.getConfiguration()
    //    VSCode.workspace.getConfiguration()
    // Because .\lib\extension.js is special, maybe it is compressed?
    //      V.default 其实等同于 VSCode
    //      V.default는 사실 VSCode와 동일합니다.
    //      V.default is actually equivalent to VSCode
    let customArg = V.default.workspace.getConfiguration().get("nsis.compiler.customArguments");
    for (let i = 0; i < customArg.length - 1; i++) {
        if (customArg[i] == "-INPUTCHARSET" && customArg[i + 1] == "UTF8") {
            console.log("用户已告知NSIS,脚本文件的编码格式为UTF8WithoutBOM");
            console.log("vscode-nsis将拦截Badtextencoding错误以防VSCode报错");
            console.log("사용자는 NSIS에 스크립트 파일의 인코딩 형식이 UTF8WithoutBOM이라고 알렸습니다.");
            console.log("vscode-nsis는 VSCode가 오류를 보고하지 않도록 Badtextencoding 오류를 차단합니다.");
            console.log("The user has informed NSIS that the encoding format of the script file is UTF8WithoutBOM");
            console.log("vscode-nsis will intercept Badtextencoding error to prevent VSCode from reporting error");

            if (result.toString().startsWith("Bad text encoding")) {
                if(firstVdefaultwindowshowInformationMessage){
                    firstVdefaultwindowshowInformationMessage = false;
                    V.default.window.showInformationMessage("时刻保证你的脚本文件真的是utf8withoutBOM格式,扩展不会一直提示你");
                    V.default.window.showInformationMessage("항상 당신의 스크립트 파일이 정말 utf8withoutBOM 형식이라는 것을 보증합니다. 확장은 항상 당신에게 제시하지 않습니다");
                    V.default.window.showInformationMessage("Always ensure that your script file is in utf8without BOM format, and the extension will not always prompt you");
                }
                return {};
            }
            // break;
        }
    }
    // -*- FINISH -*-

    if (result !== null) {
        const errorLine = parseInt(result.groups.line) - 1;

        return {
            code: '',
            message: result.groups.message,
            range: new V.default.Range(new V.default.Position(errorLine, 0), new V.default.Position(errorLine, getLineLength(errorLine))),
            severity: V.default.DiagnosticSeverity.Error
        };
    }

    return {};
}
```

**Comment by idleberg** [comment]
So you're basically suggesting a user-facing error message?

**Comment by beibeibeibei2022** [comment]
You mean the mistake on the screenshot, yes!
![image](https://user-images.githubusercontent.com/98100055/213954439-c7b74fbc-292f-42e7-a3da-5c116aa4caae.png)
When someone uses the INPUTCHARSET option, it means that the user can handle it by himself. vscode-nsis and vscode should not prompt the "bad text encoding" error.

**Comment by beibeibeibei2022** [comment]
```js
//Trick for javascript, just console.log
let customArguments = ["-INPUTCHARSETUTF8"];
let vscodeNsisCore = console.log;
let alertToUserOnce = console.log;
let userwantmodify = true;
let opensetting = console.log;

//IMPORTANT;vscode-nsis should be silent
if (customArguments.includes("-INPUTCHARSETUTF8")) {
    vscodeNsisCore(`Don't tell anyone we got a "Bad text encoding" error`);
    // User uses utf8; User is the master; NSIS、VSCode and vscode-nsis should be silent;
    alertToUserOnce(`You are using the INPUTCHARSETUTF8 option. Confirm that your file encoded by utf8withBOM! If you don't want to use it, please modify it in the settings`);
    if (userwantmodify) {
        openSetting(`Nsis › Compiler: Custom Arguments`);
    }
}
```

**Comment by beibeibeibei2022** [comment]
If someone can accurately identify the encoded utf8 without BOM file, this guy is a liar; No one can do it. This guy is just guessing with tricks;
When someone makes it clear that a file is encoded by utf8 without BOM, you should be happy because there is no need to guess and no need to find an error.

**Enable INPUTCHARSETUTF8, Disable vscode-nsis finderror.**

**Comment by beibeibeibei2022** [comment]
```
; If you want to test it, VSCode - File - New file(Ctrl+N), Copy text into this file
; Set the encoding to utf8, Save as .nsi script file.
; Hello; 你好; 안녕하세요; こんにちは
; EndIf
```

![image](https://user-images.githubusercontent.com/98100055/214202499-2953b979-b3cb-4885-95cb-dfbab178e224.png)
```diff
- VSCode-nsis +      UTF-8     = "Bad text encoding" error
+ VSCode-nsis + UTF-8 with BOM = perfect
```

**Comment by beibeibeibei2022** [comment]
## Solution (will not destroy the diagnostics feature)
1 Find file %USERPROFILE%\.vscode\extensions\idleberg.nsis-3.52.0\lib\extension.js
2 Find a function with "INPUTCHARSET" in extension.js
3 Find a function with "INPUTCHARSET" in extension.js, like 'function qn(**e**, n) {...', we need this <code>e</code>(e same as the fourth)
4 Find the line with "INPUTCHARSET" in extension.js, like '**e**.push("-INPUTCHARSET"...', we need this <code>e</code>(e same as the third)
5 Call **supportutf8 function** at the beginning of this function, use the <code>e</code> as argument
6 like this:
```js
// "Prettier - Code formatter" used, https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode
////////////🢃
function qn(e, n) {
    //////////////////////////////////////////////////////////////////////////////////////////
    ////////////🢃
    supportutf8(e); // 🢀🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜🢜
    function supportutf8(ArgList) {
        const Config = require("vscode").workspace.getConfiguration();
        const cArg = Config.get("nsis.compiler.customArguments");
        const Iarg = "INPUTCHARSET";
        const Uarg = "UTF8";
        const IARG = ["-" + Iarg, "/" + Iarg];
        for (let i = 0; i < cArg.length - 1; i++) {
            if (IARG.includes(cArg[i]) && cArg[i + 1] == Uarg) {
                ArgList.push(IARG[0], Uarg);
            }
        }
    }
    //////////////////////////////////////////////////////////////////////////////////////////
    return Te(this, void 0, void 0, function () {
        var t, a, i, r, o, s, u;
        return Ne(this, function (l) {
            switch (l.label) {
......
......
```

The variable that often use <code>.push()</code> is what we are looking for
like this
```js
e.push("-NOCD")
e.push("-NOCONFIG")
e.push("-PAUSE")
```

√ When compiling, vscode-nsis works normally.
× When diagnosing, vscode-nsis works incorrectly and discards the "- INPUTCHARSET UTF8" parameter.
√ supportutf8 fixes it.

[nsiScriptFileforTest.zip](https://github.com/idleberg/vscode-nsis/files/10543063/nsiScriptFileforTest.zip)

**Comment by idleberg** [comment]
When building Unicode installers, NSIS only supports UTF-8 with BOM and UTF-16 with BOM

**Comment by beibeibeibei2022** [comment]
In the NSIS document, it is explicitly stated that <code>UTF8 Without BOM</code> is supported.
vscode-nsis can also support <code>UTF8 Without BOM</code> with the previously mentioned [supportutf8](https://github.com/idleberg/vscode-nsis/issues/46#issuecomment-1409801283) method.

🢂[NSIS Documentation](https://nsis.sourceforge.io/Docs/Chapter3.html) 
🢃NSIS Documentation screenshot
![1](https://user-images.githubusercontent.com/98100055/216587860-6b5adbaf-ff39-4826-b1ba-38d7f73ced42.png)

A bird with wings, it's strange. <code>UTF-16 with BOM</code>, it's strange too. No one use <code>UTF-16 without BOM</code>.
UTF16 family has three brothers, <code>UTF-16</code>/<code>UTF-16BE</code>/<code>UTF-16LE</code>, They all have BOM.
Although NSIS supports <code>UTF-16 without BOM</code>, but nobody cares.

UTF8 can live well without BOM. BOM is redundant.<code>UTF8 without BOM</code> is important.

Again，I tracked the parameters.
When vscode-nsis know VSCode open a .nsi file, vscode-nsis **NOT** use INPUTCHARSET.
When vscode-nsis compiling, vscode-nsis use INPUTCHARSET.
When vscode-nsis know VSCode save a .nsi file, vscode-nsis diagnosing, vscode-nsis **NOT** use INPUTCHARSET.("-XOutFile NUL" will be used when diagnosing)
When vscode-nsis get Command Reference or show version or show compiler flags, vscode-nsis not use INPUTCHARSET.(only -CMDHELP or -VERSION or -HDRINFO), This situation is all right.
"Nsis › Compiler: Custom Arguments", this setting item is unstable. It is normal in some cases and lost in some cases.

When you say "only supports", You forget that vscode-nsis can use INPUTCHARSET and work normally(compiling).
If you are right, then #31 is not solved, #31 also uses utf8 <code>-INPUTCHARSET UTF8</code>, just like me.
#31 screenshot
![image](https://user-images.githubusercontent.com/98100055/216616746-602e12bc-3b93-4b70-85d2-c7c7be19d8eb.png)
Use the previously mentioned [supportutf8](https://github.com/idleberg/vscode-nsis/issues/46#issuecomment-1409801283) method,  "한글_테스트.exe" red line can be solved in #31 screenshot.

**Comment by beibeibeibei2022** [comment]
@idleberg use [supportutf8](https://github.com/idleberg/vscode-nsis/issues/46#issuecomment-1409801283) can solve the problem, which has nothing to do with nsis.
